spec/requests/comments_create_spec.rb

```diff
@@ -4,27 +4,36 @@
   let(:user) { create(:user) }
   let(:blocker) { create(:user) }
   let(:article) { create(:article, user_id: user.id) }
+  let(:new_body) { -> { "NEW BODY #{rand(100)}" } }
 
   before do
     sign_in user
   end
 
   it "creates ordinary article with proper params" do
-    new_body = "NEW BODY #{rand(100)}"
     post "/comments", params: {
-      comment: { body_markdown: new_body, commentable_id: article.id, commentable_type: "Article" }
+      comment: { body_markdown: new_body.call, commentable_id: article.id, commentable_type: "Article" }
     }
     expect(Comment.last.user_id).to eq(user.id)
   end
 
   it "creates NotificationSubscription for comment" do
-    new_body = "NEW BODY #{rand(100)}"
     post "/comments", params: {
-      comment: { body_markdown: new_body, commentable_id: article.id, commentable_type: "Article" }
+      comment: { body_markdown: new_body.call, commentable_id: article.id, commentable_type: "Article" }
     }
     expect(NotificationSubscription.last.notifiable).to eq(Comment.last)
   end
 
+  it "returns 429 Too Many Requests when a user reachers their rate limit" do
+    create_list(:comment, 10, user: user, commentable: article)
```

Comment:
It would be nice to stub `RateLimitChecker` response instead of creating 10 comments records, though that's not easy in the current implementation :thinking:

Reply:
Started with stubbing, reached same conclusion as you, decided to just create comments instead.

Reply:
@lightalloy To clarify, we can of course do the following:

```ruby
allow_any_instance_of(RateLimitChecker).to receive(:limit_by_action).with("comment_creation").and_return(true)
```

But the RSpec docs consider it an antipattern and so do I. To quote [the docs](https://relishapp.com/rspec/rspec-mocks/docs/working-with-legacy-code/any-instance):

```
This feature is sometimes useful when working with legacy code, though in general we
discourage its use for a number of reasons:

* The rspec-mocks API is designed for individual object instances, but this feature operates on entire classes of objects. As a result there are some semantically confusing edge cases. For example, in expect_any_instance_of(Widget).to receive(:name).twice it isn't clear whether a specific instance is expected to receive name twice, or if two receives total are expected. (It's the former.)

* Using this feature is often a design smell. It may be that your test is trying to do too much or that the object under test is too complex.

* It is the most complicated feature of rspec-mocks, and has historically received the most bug reports. (None of the core team actively use it, which doesn't help.)
```

Reply:
That said, I go with what most people feel comfortable with, any opinions @rhymes or @mstruve?

Reply:
In an ideal situation RateLimitChecker would be configurable, I really don't like that limits are hardcoded but we aren't there yet. Maybe we can do that in a separate PR? 

Though a question arises: should we refactor classes, even if not great, only for testing purposes?

Reply:
> Though a question arises: should we refactor classes, even if not great, only for testing purposes?

IMO no, classes are written to execute code and doing that well and in an organized fashion should be their first priority. I will concede though there are times where not being able to test something easily exposes flaws in the code's ability to do its primary function well but in this case I think we are fine.

Reply:
I agree with not using `allow_any_instance_of`, I have other ideas but they require design changes which I'm also not very comfortable with (and they shouldn't be part of this pr anyway), so I agree on keeping creating comments logic here.

Reply:
@lightalloy Yup, everything else I tried would have necessitated some changes to `RateLimitChecker` that felt too far outside the scope of this PR.

In the general discussion, I tend to agree with @mstruve on this. *If* there are two similar ways of achieving things, but one is much better for testing, by all means go for that one. But obscuring core functionality or introducing something that's otherwise not needed (DI where a non-default dependency is only injected in tests) is not a good path to go down on in my opinion.

Reply:
In my opinion, often it makes sense to adjust the code to be testable, including injecting dependency. Of course, every such decision should be thought on.